File: packages/page/src/component-walk.ts
/**
 * 页面组件树遍历:内容分区顶层组件,以及 Tab 容器内的子组件。
 * 校验、能力探测、编排与导航都走这一处,避免各层各写一遍漏掉嵌套表。
 *
 * 只依赖 `type` 与可选 `props.tabs`,不依赖 `./page` 的领域类型,以免成环。
 */

type Json = Record<string, unknown>;

export function walkComponents<T extends { type: string }>(
  components: readonly T[],
  basePath: string,
  visit: (component: T, path: string) => void
): void {
  components.forEach((component, index) => {
    visitTree(component, `${basePath}/${index}`, visit);
  });
}

export function walkPageComponents<T extends { type: string }>(
  page: { sections: ReadonlyArray<{ components: readonly T[] }> },
  visit: (component: T, path: string) => void
): void {
  page.sections.forEach((section, sectionIndex) => {
    walkComponents(section.components, `/sections/${sectionIndex}/components`, visit);
  });
}

export function flattenPageComponents<T extends { type: string }>(
  page: { sections: ReadonlyArray<{ components: readonly T[] }> }
): T[] {
  const components: T[] = [];
  walkPageComponents(page, (component) => {
    components.push(component);
  });
  return components;
}

function visitTree<T extends { type: string }>(
  component: T,
  path: string,
  visit: (component: T, path: string) => void
): void {
  visit(component, path);
  if (component.type !== 'tabContainer') return;
  const tabs = (component as { props?: { tabs?: Array<{ component?: T }> } }).props?.tabs ?? [];
  tabs.forEach((tab, tabIndex) => {
    if (!tab.component) return;
    visitTree(tab.component, `${path}/props/tabs/${tabIndex}/component`, visit);
  });
}

function record(value: unknown): Json | undefined {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
    ? (value as Json)
    : undefined;
}

/**
 * 原始文档上的组件树遍历,供能力下限探测使用。
 * 必须读原始文档:解析接缝会把文本取值引用整值替换掉。
 */
export function walkDocumentComponents(
  document: unknown,
  visit: (component: Json, path: string) => void
): void {
  const sections = record(document)?.sections;
  if (!Array.isArray(sections)) return;
  sections.forEach((sectionCandidate, sectionIndex) => {
    const components = record(sectionCandidate)?.components;
    if (!Array.isArray(components)) return;
    components.forEach((child, componentIndex) => {
      visitDocumentTree(
        child,
        `/sections/${sectionIndex}/components/${componentIndex}`,
        visit
      );
    });
  });
}

function visitDocumentTree(
  candidate: unknown,
  path: string,
  visit: (component: Json, path: string) => void
): void {
  const component = record(candidate);
  if (!component) return;
  visit(component, path);
  if (component.type !== 'tabContainer') return;
  const tabs = record(component.props)?.tabs;
  if (!Array.isArray(tabs)) return;
  tabs.forEach((tabCandidate, tabIndex) => {
    visitDocumentTree(
      record(tabCandidate)?.component,
      `${path}/props/tabs/${tabIndex}/component`,
      visit
    );
  });
}

File: metriccanvas-page-assets/page-assets-service/src/main/java/com/huawei/cdi/pageassets/domain/page/document/ComponentWalk.java
package com.huawei.cdi.pageassets.domain.page.document;

import com.fasterxml.jackson.databind.JsonNode;
import com.huawei.cdi.pageassets.domain.page.json.Json;

import java.util.function.BiConsumer;

/**
 * 页面组件树遍历：内容分区顶层组件，以及两种容器内的子组件——Tab 容器的
 * {@code props.tabs[].component / components[]} 与组合卡的 {@code props.components[]}。
 * 校验与能力探测都走这一处；既可遍历原始文档，也可遍历解析产物（形状相同）。
 */
public final class ComponentWalk {
    private ComponentWalk() {
    }

    public static void walkDocument(JsonNode document, BiConsumer<JsonNode, String> visit) {
        JsonNode sections = Json.get(document, "sections");
        if (!Json.isArray(sections)) {
            return;
        }
        for (int sectionIndex = 0; sectionIndex < sections.size(); sectionIndex++) {
            walkSection(sections.get(sectionIndex), "/sections/" + sectionIndex + "/components", visit);
        }
    }

    public static void walkSection(JsonNode section, String basePath, BiConsumer<JsonNode, String> visit) {
        JsonNode components = Json.get(section, "components");
        walkComponents(components, basePath, visit);
    }

    public static void walkComponents(JsonNode components, String basePath, BiConsumer<JsonNode, String> visit) {
        if (!Json.isArray(components)) {
            return;
        }
        for (int index = 0; index < components.size(); index++) {
            visitTree(components.get(index), basePath + "/" + index, visit);
        }
    }

    private static void visitTree(JsonNode candidate, String path, BiConsumer<JsonNode, String> visit) {
        JsonNode component = Json.record(candidate);
        if (component == null) {
            return;
        }
        visit.accept(component, path);
        String type = Json.text(component.get("type"));
        JsonNode props = Json.record(component.get("props"));
        if ("tabContainer".equals(type)) {
            JsonNode tabs = Json.get(props, "tabs");
            if (!Json.isArray(tabs)) {
                return;
            }
            for (int tabIndex = 0; tabIndex < tabs.size(); tabIndex++) {
                JsonNode tab = Json.record(tabs.get(tabIndex));
                JsonNode single = Json.get(tab, "component");
                if (single != null && !single.isNull() && truthy(single)) {
                    visitTree(single, path + "/props/tabs/" + tabIndex + "/component", visit);
                    continue;
                }
                JsonNode children = Json.get(tab, "components");
                if (!Json.isArray(children)) {
                    continue;
                }
                for (int childIndex = 0; childIndex < children.size(); childIndex++) {
                    visitTree(children.get(childIndex),
                            path + "/props/tabs/" + tabIndex + "/components/" + childIndex, visit);
                }
            }
            return;
        }
        if ("compositeCard".equals(type)) {
            JsonNode children = Json.get(props, "components");
            if (!Json.isArray(children)) {
                return;
            }
            for (int childIndex = 0; childIndex < children.size(); childIndex++) {
                visitTree(children.get(childIndex), path + "/props/components/" + childIndex, visit);
            }
        }
    }

    /** JS 真值判断，只在 `tab.component` 这一处需要（基线写的是 `if (tab?.component)`）。 */
    private static boolean truthy(JsonNode node) {
        if (node.isBoolean()) {
            return node.booleanValue();
        }
        if (node.isNumber()) {
            return node.decimalValue().signum() != 0;
        }
        if (node.isTextual()) {
            return !node.textValue().isEmpty();
        }
        return !node.isNull();
    }
}
